sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala

```diff
@@ -444,6 +447,156 @@ case class UnresolvedStar(target: Option[Seq[String]]) extends Star with Unevalu
+    //            struct fields in order to rewrite the struct
+    //       ii.  a remainingExcept is empty, but there are multiple remainingExcepts => we must
+    //            have duplicate/overlapping excepts - throw an error
+    //       iii. [otherwise] remainingExcept is exactly Seq(Seq()) => this is the base 'filtering'
+    //            case. we omit the column from the output (this is a column we would like to
+    //            except). NOTE: this case isn't explicitly listed in the `collect` below since we
+    //            'collect' columns which match the cases above and omit ones that fall into this
+    //            remaining case.
+    def filterColumns(columns: Seq[NamedExpression], excepts: Seq[(NamedExpression, Seq[String])])
+      : Seq[NamedExpression] = {
+      // group the except pairs by the column they refer to. NOTE: no groupMap until scala 2.13
+      val groupedExcepts: AttributeMap[Seq[Seq[String]]] =
+        AttributeMap(excepts.groupBy(_._1.toAttribute).mapValues(v => v.map(_._2)))
+
+      // map input columns while searching for the except entry corresponding to the current column
+      columns.map(col => col -> groupedExcepts.get(col.toAttribute)).collect {
+        // pass through columns that don't match anything in groupedExcepts
+        case (col, None) => col
+        // found a match but nestedExcepts has remaining excepts - recurse to rewrite the struct
+        case (col, Some(nestedExcepts)) if !nestedExcepts.exists(_.isEmpty) =>
+          val fields = col.dataType match {
+            case s: StructType => s.fields
+            // we shouldn't be here since we throw the same error above (in getRootStruct), but
+            // nonetheless just throw the same error
+            case _ => throw new AnalysisException(
+              errorClass = "EXCEPT_NESTED_COLUMN_INVALID_TYPE",
+              messageParameters =
+                Map("columnName" -> col.qualifiedName, "dataType" -> col.dataType.toString))
+          }
+          val extractedFields = fields.zipWithIndex.map { case (f, i) =>
+            Alias(GetStructField(col, i), f.name)()
+          }
+          val newExcepts = nestedExcepts.map { nestedExcept =>
+            // INVARIANT: we cannot have duplicate column names in nested columns, thus, this `head`
+            // will find the one and only column corresponding to the correct extractedField.
+            extractedFields.collectFirst { case col if resolver(col.name, nestedExcept.head) =>
+              col.toAttribute -> nestedExcept.tail
+            }.get
+          }
+          Alias(CreateStruct(filterColumns(extractedFields.toSeq, newExcepts)), col.name)()
+        // if there are multiple nestedExcepts but one is empty we must have overlapping except
+        // columns. throw an error.
+        case (col, Some(nestedExcepts)) if nestedExcepts.size > 1 =>
+          throw new AnalysisException(errorClass = "EXCEPT_OVERLAPPING_COLUMNS", Map.empty)
```

Comment:
it's not overlapping but duplicated?

Reply:
overlapping is more accurate as this error is thrown even with overlapping columns in nested structs - e.g.
select * except(data.a.b, data.a) should throw this exception as well

Reply:
oh I see!